### PR TITLE
SmartField migration to typescript (rename & add types)

### PR DESCRIPTION
**Pull request by fschinkel**

### CODE REVIEW

File: eclipse-scout-core/src/form/fields/smartfield/ProposalChooser.ts [hunk 1]
```diff
@@ -35,21 +50,21 @@ export default class ProposalChooser extends Widget {
...
     }
 
-    this.model = this._createModel();
+    this.content = this._createContent();
   }
 
-  _createModel($parent) {
-    throw new Error('_createModel() not implemented');
+  protected _createContent(): TContent {
+    throw new Error('_createContent() not implemented');
```
Comment (mvilliger) [inline eclipse-scout-core/src/form/fields/smartfield/ProposalChooser.ts]: Maybe make this method abstract?

File: eclipse-scout-core/src/form/fields/smartfield/ProposalChooser.ts [hunk 2]
```diff
@@ -35,21 +50,21 @@ export default class ProposalChooser extends Widget {
...
   /**
    * Creates a layout resetter that is used by the {@link ProposalChooserLayout}.
    */
-  _createLayoutResetter() {
+  protected _createLayoutResetter(): ProposalChooserLayoutResetter {
     throw new Error('_createLayoutResetter() not implemented');
   }
 
-  setLookupResult(result) {
+  setLookupResult(result: SmartFieldLookupResult<TValue>) {
```
Comment (mvilliger) [inline eclipse-scout-core/src/form/fields/smartfield/ProposalChooser.ts]: Maybe make this method abstract?

File: eclipse-scout-core/src/form/fields/smartfield/ProposalChooser.ts [hunk 3]
```diff
@@ -82,18 +97,22 @@ export default class ProposalChooser extends Widget {
    * Implement this function to get the selected row or node from the proposal chooser.
    * The implementation depends on the widget used by the chooser (Table or Tree).
    */
-  selectedRow() {
+  selectedRow(): TContentRow {
```
Comment (mvilliger) [inline eclipse-scout-core/src/form/fields/smartfield/ProposalChooser.ts]: Maybe make this method abstract?

File: eclipse-scout-core/src/form/fields/smartfield/ProposalChooser.ts [hunk 4]
```diff
@@ -82,18 +97,22 @@ export default class ProposalChooser extends Widget {
...
     throw new Error('selectedRow() not implemented');
   }
 
-  _render() {
+  getSelectedLookupRow(): LookupRow<TValue> {
+    throw new Error('getSelectedLookupRow() not implemented');
+  }
```
Comment (mvilliger) [inline eclipse-scout-core/src/form/fields/smartfield/ProposalChooser.ts]: Maybe make this method abstract?

File: eclipse-scout-core/src/form/fields/smartfield/ProposalChooser.ts [hunk 5]
```diff
@@ -272,4 +291,25 @@ export default class ProposalChooser extends Widget {
       this.selectFirstLookupRow();
     }
   }
+
+  trySelectCurrentValue() {
+    throw new Error('trySelectCurrentValue() not implemented');
+  }
```
Comment (mvilliger) [inline eclipse-scout-core/src/form/fields/smartfield/ProposalChooser.ts]: Maybe make this method abstract?

File: eclipse-scout-core/src/form/fields/smartfield/ProposalChooserModel.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     BSI Business Systems Integration AG - initial API and implementation
+ */
+import {SmartField, Status, WidgetModel} from '../../../index';
+
+export default interface ProposalChooserModel<TValue> extends WidgetModel {
+  smartField?: SmartField<TValue>;
+  status?: Status;
```
Comment (mvilliger) [inline eclipse-scout-core/src/form/fields/smartfield/ProposalChooserModel.ts]: StatusOrModel?

File: eclipse-scout-core/src/form/fields/smartfield/ProposalField.ts [hunk 7]
```diff
@@ -128,38 +135,29 @@ export default class ProposalField extends SmartField {
...
   }
 
-  /**
-   * This function is overridden by ProposalField because it has a different behavior than the smart-field.
-   */
-  _acceptLookupRowAndValueFromField(otherField) {
-    if (this.lookupRow !== otherField.lookupRow) {
-      this.setLookupRow(otherField.lookupRow);
-    }
-  }
```
Comment (mvilliger) [inline eclipse-scout-core/src/form/fields/smartfield/ProposalField.ts:156]: Why can this be deleted?

File: eclipse-scout-core/src/form/fields/smartfield/ProposalTreeNode.ts [hunk 8]
```diff
@@ -43,20 +45,20 @@ export default class ProposalTreeNode extends TreeNode {
     this.$node.toggleClass('inactive', !this.lookupRow.active);
   }
 
-  isBrowseLoadIncremental() {
+  isBrowseLoadIncremental(): boolean {
     return this.proposalChooser.isBrowseLoadIncremental();
   }
 
-  loadChildren() {
+  override loadChildren(): JQuery.Deferred<any> {
```
Comment (mvilliger) [inline eclipse-scout-core/src/form/fields/smartfield/ProposalTreeNode.ts]: The deferred cannot be typed?

File: eclipse-scout-core/src/form/fields/smartfield/SmartField.ts [hunk 9]
```diff
@@ -9,12 +9,69 @@
...
+  declare model: SmartFieldModel<TValue>;
+  declare eventMap: SmartFieldEventMap<TValue>;
+
+  popup: SmartFieldTouchPopup<TValue> | SmartFieldPopup<TValue>;
+  lookupCall: LookupCall<TValue>;
+  codeType: string;
+  lookupRow: LookupRow<TValue>;
+  browseHierarchy: boolean;
+
+  /** a positive number, _not_ null or undefined! */
```
Comment (mvilliger) [inline eclipse-scout-core/src/form/fields/smartfield/SmartField.ts]: jsdoc can be removed?

File: eclipse-scout-core/src/form/fields/smartfield/ProposalField.ts (hunk 7 above)
Comment (fschinkel) [reply eclipse-scout-core/src/form/fields/smartfield/ProposalField.ts:156]: The method was never called.

**Pull request merged**